### DISCUSSION TITLE
Issue 1235

### DISCUSSION THREAD

**Discussion by m-holger** [qpdf/qpdf-dev#9] [General]
I have had a very brief look at this and my initial thoughts are:

- validation of CLI options around encryption looks incomplete. It would make sense to develop a matrix of incompatible options and strengthen validations.

- validations are usually done in QPDFJob::Config or QPDFJob::checkConfiguration. For encrypted files with empty user password, some of the validation is not possible until the primary input has been opened. It would make sense to have a second round of validations at this point. I am not sure about a third round at the start of QPDFJob::writeQPDF - arguably any inconsistencies introduced in between the call to createQPDF and writeQPDF are the programmers responsibility and the checks in QPDFWriter are sufficient.

- as I understand it, encryption only uses the first part of the id. Without thinking this fully through, it would look possible to only generate the id when setting up encryption if part 1 can't be found, which would allow deterministic-id to be used with encrypted files with existing id.

**Comment by jberkenbilt** [comment]
> * validation of CLI options around encryption looks incomplete. It would make sense to develop a matrix of incompatible options and strengthen validations.

Probably true. My original intention with most qpdf options is that, if there is a contradiction, later ones override earlier ones, but I no longer agree with that approach at least in most cases. My original reasoning was that it enabled you to create an alias and then override things, but it's too sloppy, and I've evolved away from that thinking.

If both options are given from the library by directly calling QPDFWriter options, it should be an exception (probably even a logic error), but not an INTERNAL ERROR. That's just incorrect. A logic error is appropriate if qpdf never calls it that way itself (by adding checks to QPDFJob) since it would indicate incorrect use of the library by the caller.

> * validations are usually done in QPDFJob::Config or QPDFJob::checkConfiguration. For encrypted files with empty user password, some of the validation is not possible until the primary input has been opened. It would make sense to have a second round of validations at this point. I am not sure about a third round at the start of QPDFJob::writeQPDF - arguably any inconsistencies introduced in between the call to createQPDF and writeQPDF are the programmers responsibility and the checks in QPDFWriter are sufficient.
 
Yeah, maybe that's why I didn't implement that check, but I agree -- by the time we are ready to configure QPDFWriter, we should know whether the output is supposed to be encrypted.

> * as I understand it, encryption only uses the first part of the id. Without thinking this fully through, it would look possible to only generate the id when setting up encryption if part 1 can't be found, which would allow deterministic-id to be used with encrypted files with existing id.

Perhaps. It's been a long time since I looked at that code. There may be other solutions as well. I think it's entirely reasonable to simply disallow deterministic ID with encrypted files. If you use AES256, there's a random block, and encrypted output is never consistent, so having a deterministic ID is useless.

**Comment by jberkenbilt** [comment]
If we are going to discuss fixes to regular issues here (which I don't think we should, but I'm fine letting you make the call on that), we need to link back to the original issue at least. In this case, it's qpdf/qpdf#1235.
